File: csir_intelligent_manipulator/csir_wam_object_manipulation/bhand_grasp_controller/test/utest_bhand_grasp_controller.py
#! /usr/bin/env python

# Copyright (c) 2014 CSIR. All rights reserved.
# Use of this source code is governed by the MIT license that can be found in
# the LICENSE file.
# Author: Jimmy Kizito

""" Unit tests for bhand_grasp_controller

Classes:
    BHandGraspControllerUTest: unit test fixture
"""

PKG = "bhand_grasp_controller"
import roslib
roslib.load_manifest(PKG)
import unittest
from bhand_grasp_controller import *

class BHandGraspControllerUTest(unittest.TestCase):
    """ Unit test fixture for testing bhand_grasp_controller.

    Attributes:
        _bgc: BHandGraspController class under test
        _bhand_joints: list of state (in radians) of 3 proximal finger joints.
            Varies from 0 to 140 degrees.
    """

    def setUp(self):
        """ Initialise class members for test.
        """
        self._bgc = BHandGraspController()
        self._bhand_joints = [0.0] * BHandGraspController._DOF

    def test_convert_joints_to_grasp(self):
        """ Test convert_joints_to_grasp function.

        Test that input values yield expected values and that and that "small"
        changes in the input are ignored.
        """
        # Expect disturbance invariant output
        expected_outputs = [0.0, 0.5, 1.0]
        for i in expected_outputs:
            theta = BHandGraspController._MAX_RANGE * i;
            # [-tol, 0, tol]
            for tol in [j * BHandGraspController._JOINT_TOL for j in range(-1, 2)]:
                self._bhand_joints =  [theta + tol] * len(self._bhand_joints)
                grasp_metric =\
                    self._bgc.convert_joints_to_grasp(self._bhand_joints)
                self.assertAlmostEqual(grasp_metric, i, self._bgc._decimal_places)
        # Check deviations around mid-point
        theta = 0.5 * BHandGraspController._MAX_RANGE
        self._bhand_joints =  [theta] * len(self._bhand_joints);
        self._bhand_joints[1] = 0.25 * BHandGraspController._MAX_RANGE
        grasp_metric = self._bgc.convert_joints_to_grasp(self._bhand_joints)
        self.assertLess(grasp_metric, 0.5)
        self.assertGreater(grasp_metric, 0.0)
        self._bhand_joints[1] = 0.75 * BHandGraspController._MAX_RANGE
        grasp_metric = self._bgc.convert_joints_to_grasp(self._bhand_joints)
        self.assertLess(grasp_metric, 1.0)
        self.assertGreater(grasp_metric, 0.5)

if __name__ == "__main__":
    import rosunit
    rosunit.unitrun(PKG, "utest_bhand_grasp_controller",
                    BHandGraspControllerUTest)

File: csir_intelligent_manipulator/csir_wam_object_manipulation/bhand_grasp_controller/test/utest_bhand_grasp_controller.cpp
/*
 * Copyright (c) 2014 CSIR. All rights reserved.
 * Use of this source code is governed by the MIT license that can be found in
 * the LICENSE file.
 */

#include "bhand_grasp_controller/bhand_grasp_controller.h"
//#include <boost/scoped_ptr.hpp>
#include <vector>
#include <gtest/gtest.h>

/*!
 * @brief Unit test fixture for BHandGraspController
 */
class BHandGraspControllerUTest : public ::testing::Test
{
 public:
  /*!
   * @brief The constructor serves only to initialise members of the test fixture
   * class.
   */
  BHandGraspControllerUTest() : bHandJoints_(BHandGraspController::DOF_, 0.0)
  {
    graspTol = (BHandGraspController::JOINT_TOL_ /
                BHandGraspController::MAX_RANGE_);
  }

 protected:
  //virtual void SetUp()
  //virtual void TearDown()

  /*! Class under test. */
  BHandGraspController bgc_;
  /*! Vector of state (in radians) of 3 proximal finger joints. */
  std::vector<double> bHandJoints_;
  double graspTol;
};

/*!
 * @brief Test convertJointsToGrasp function.
 * @details Test that input values yield expected values and that "small" changes
 * in the input are ignored.
 */
TEST_F(BHandGraspControllerUTest, convertJointsToGrasp)
{
  double theta = 0.0;
  double graspMetric = 0.0;
  // Expect disturbance invariant output
  for (double expectedOutput = 0.0; expectedOutput < 1.0; expectedOutput += 0.5)
  {
    theta = BHandGraspController::MAX_RANGE_ * expectedOutput;
    ROS_DEBUG("theta: %f", theta);
    for (double tol = -BHandGraspController::JOINT_TOL_; tol < BHandGraspController::JOINT_TOL_ + tol/10.0;
         tol += BHandGraspController::JOINT_TOL_)
    {
      bHandJoints_.assign(bHandJoints_.size(), theta + tol);
      ROS_DEBUG("bhj: %f, %f, %f", bHandJoints_.at(0), bHandJoints_.at(1), bHandJoints_.at(2));
      graspMetric = bgc_.convertJointsToGrasp(bHandJoints_);
      EXPECT_NEAR(graspMetric, expectedOutput, graspTol) <<
          "graspMetric too sensitive to joint variance";
    }
  }
  // Check deviations around range
  theta = 0.5 * BHandGraspController::MAX_RANGE_;
  bHandJoints_.assign(bHandJoints_.size(), theta);
  bHandJoints_.at(1) = 0.25 * BHandGraspController::MAX_RANGE_;
  graspMetric = bgc_.convertJointsToGrasp(bHandJoints_);
  EXPECT_LT(graspMetric, 0.5);
  EXPECT_GT(graspMetric, 0.0);
  bHandJoints_.at(1) = 0.75 * BHandGraspController::MAX_RANGE_;
  graspMetric = bgc_.convertJointsToGrasp(bHandJoints_);
  EXPECT_LT(graspMetric, 1.0);
  EXPECT_GT(graspMetric, 0.5);
}

int main(int argc, char **argv)
{
  ::testing::InitGoogleTest(&argc, argv);
  ros::init(argc, argv, "utest_bhand_grasp_controller");
  return RUN_ALL_TESTS();
}
